### PR TITLE
Recipe: Reinforcement learning for generative models using FSDP as Ba…

### PR DESCRIPTION
**Pull request by wwwbby**
### summary
This RFC proposes adding a comprehensive recipe to the verl-recipe repository that demonstrates how to integrate the verl reinforcement learning framework with [FSDP](https://docs.pytorch.ac.cn/tutorials/intermediate/FSDP1_tutorial.html) (Ascend Multi-modal Model Suite) to implement RLHF (Reinforcement Learning from Human Feedback) training for generative models on NPU (Ascend AI Processors). The recipe will feature the [DanceGRPO](https://dancegrpo.github.io/) algorithm, using [mammothmoda](https://github.com/bytedance/mammothmoda/tree/main) as the policy model and [HPSv3](https://github.com/MizzenAI/HPSv3) as the reward model, providing an end-to-end training workflow.
### result
image001

### CODE REVIEW

File: dance_grpo/dance_grpo_fsdp/README.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
+<p align="center">
+<h1 align="center"> Recipe: Reinforcement learning for generative models using FSDP as the backend (DanceGRPO) </h1>
+
+## 1. Environment installation ##
+
```
Comment (wwwbby) [inline dance_grpo/dance_grpo_fsdp/README.md:5]: 补充CANN版本，torch版本

File: dance_grpo/dance_grpo_fsdp/README.md [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+# verl-recipe
+# verl
+# ├── recipe
+#     ├── dance_grpo
+#         ├── dance_grpo_fsdp
+```
+
+## 2. Dataset preparation ##
+
+Reference ` verl/recipe/dance_grpo/dance_grpo_fsdp/data/prompt.json ` In the example provided in, you can replace the customized prompt text and run the following command to generate a parquet file:
```
Comment (wwwbby) [inline dance_grpo/dance_grpo_fsdp/README.md:54]: 贴数据集hf链接

File: dance_grpo/dance_grpo_fsdp/README.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+| `--config-path` | `config`            | Hydra configuration directory |
+| `--config-name` | `dance_ppo_trainer` | Hydra configuration filename  |
+
+Start RL training.
+
+```shell
+# cd verl
+
+bash recipe/dance_grpo/dance_grpo_fsdp/run_verl_dance.sh
+```
```
Comment (wwwbby) [inline dance_grpo/dance_grpo_fsdp/README.md:199]: 添加效果图片

File: dance_grpo/dance_grpo_fsdp/dance_ray_trainer.py [hunk 4]
```diff
@@ -0,0 +1,616 @@
...
+        if self.config.actor_rollout_ref.actor.strategy in {"fsdp", "fsdp2"}:
+            resource_pool = self.resource_pool_manager.get_resource_pool(actor_role)
+            actor_rollout_cls = RayClassWithInitArgs(
+                cls=self.role_worker_mapping[actor_role],
+                config=self.config.actor_rollout_ref,
+                role=str(actor_role),
+            )
+            self.resource_pool_to_cls[resource_pool][str(actor_role)] = actor_rollout_cls
+
+            # Create Mllm worker
```
Comment (wwwbby) [inline dance_grpo/dance_grpo_fsdp/dance_ray_trainer.py:186]: 删掉注释代码

File: dance_grpo/dance_grpo_fsdp/dance_ray_trainer.py [hunk 5]
```diff
@@ -0,0 +1,616 @@
...
+                assert str(Role.ActorRolloutRef) in all_wg, f"{all_wg.keys()=}"
+                self.ref_policy_wg = all_wg[str(Role.ActorRolloutRef)]
+
+        self.rm_wg = None
+        # initalization of rm_wg will be deprecated in the future
+        if self.use_rm:
+            self.rm_wg = all_wg[str(Role.RewardModel)]
+            self.rm_wg.init_model()
+
+        # we should create rollout at the end so that vllm can have a better estimation of kv cache memory
```
Comment (wwwbby) [inline dance_grpo/dance_grpo_fsdp/dance_ray_trainer.py:273]: 删掉被注释的代码

File: dance_grpo/dance_grpo_fsdp/dance_ray_trainer.py [hunk 6]
```diff
@@ -0,0 +1,616 @@
...
+                    k_partitions=world_size,
+                    equal_size=True,
+                )
+                for j, part in enumerate(rearrange_minibatch_lst):
+                    global_partition_lst[j].extend([x + minibatch_size * i for x in part])
+        else:
+            global_partition_lst = get_seqlen_balanced_partitions(
+                workload_lst, k_partitions=world_size, equal_size=True
+            )
+        # # Place smaller micro-batches at both ends to reduce the bubbles in pipeline parallel.
```
Comment (wwwbby) [inline dance_grpo/dance_grpo_fsdp/dance_ray_trainer.py:328]: 有两个注释符

File: dance_grpo/dance_grpo_fsdp/main_dance.py [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+
+def create_rl_dataset(data_paths, data_config, tokenizer, processor, is_train=True, max_samples: int = -1):
+    """Create a dataset that parses pickapic JSON structure and extracts only the prompt field."""
+    import json
+    import os
+
+    import numpy as np
+    from torch.utils.data import Dataset
+
+    class PickAPicDataset(Dataset):
```
Comment (wwwbby) [inline dance_grpo/dance_grpo_fsdp/main_dance.py:296]: 建议放在dataset文件里

File: dance_grpo/dance_grpo_fsdp/diffusion_workers.py [hunk 8]
```diff
@@ -0,0 +1,1380 @@
...
+        rollout_config = self.config.rollout
+        actor_config = self.config.actor
+        self.rollout = HFRollout(mllm, dit, vae, full_model, self.processor, config=rollout_config)
+        self.actor = DataParallelPPOActor(self.dit, actor_config)
+        self.inferencer = HPSv3RewardInferencer(device="cpu", checkpoint_path=self.config.model.reward_model_path)
+        self.inferencer.model.eval()
+        for param in self.inferencer.model.parameters():
+            param.requires_grad = False
+
+        # Initialize checkpoint manager
```
Comment (LLMShark) [inline dance_grpo/dance_grpo_fsdp/diffusion_workers.py:339]: 注释代码

File: dance_grpo/dance_grpo_fsdp/diffusion_workers.py [hunk 9]
```diff
@@ -0,0 +1,1380 @@
...
+        data.batch = output.batch
+        return data.union(output)
+
+    def compute_rewards(self, data: DataProto):
+        """Compute rewards for the generated sequences"""
+        prompts = data.non_tensor_batch["prompts"].tolist()
+        all_images = data.non_tensor_batch["all_images"]
+        all_images_pil = [Image.fromarray(img_np) for img_np in all_images]
+
+        def get_temp_paths(images):
```
Comment (LLMShark) [inline dance_grpo/dance_grpo_fsdp/diffusion_workers.py:409]: get_temp_paths这些辅助函数可以优化掉

File: dance_grpo/dance_grpo_fsdp/diffusion_workers.py [hunk 10]
```diff
@@ -0,0 +1,1380 @@
...
+        # Apply monkey patches
+        # Setup actor model with FSDP
+        log_gpu_memory_usage("Before init_fsdp_module", logger=logger, level=logging.INFO)
+
+        mllm, dit, vae, full_model, processor = self.extract_and_wrap_models(self.config.model.path)
+
+        if torch.distributed.get_rank() == 0:
+            logger.info("Successfully loaded model.")
+
+        self.dit = dit
```
Comment (LLMShark) [inline dance_grpo/dance_grpo_fsdp/diffusion_workers.py:300]: 审视一下actor.py是否还需要

File: dance_grpo/dance_grpo_fsdp/diffusion_workers.py [hunk 11]
```diff
@@ -0,0 +1,1380 @@
...
+            num_warmup_steps=self.config.actor.optim.lr_scheduler.num_warmup_steps,
+            num_training_steps=self.config.actor.optim.lr_scheduler.num_training_steps,
+            num_cycles=self.config.actor.optim.lr_scheduler.get("num_cycles", 1),
+            power=self.config.actor.optim.lr_scheduler.get("power", 1.0),
+        )
+
+        # Create a custom config for HFRollout that includes diffusion-specific parameters
+        rollout_config = self.config.rollout
+        actor_config = self.config.actor
+        self.rollout = HFRollout(mllm, dit, vae, full_model, self.processor, config=rollout_config)
```
Comment (LLMShark) [inline dance_grpo/dance_grpo_fsdp/diffusion_workers.py:332]: 审视一下，rollout/actor包装起来，不要都暴露在worker实现里

File: dance_grpo/dance_grpo_fsdp/diffusion_workers.py [hunk 12]
```diff
@@ -0,0 +1,1380 @@
...
+            )
+
+            # -------------------------------------------------------------- #
+            # 5.  Compute GRPO advantages per group                           #
+            # -------------------------------------------------------------- #
+            batch_size = all_latents.shape[0]
+            advantages = torch.zeros(batch_size, device=device, dtype=torch.float32)
+            group_indices = torch.chunk(torch.arange(batch_size, device=device), batch_size // grpo_size)
+            group_reward_stds = []
+            for group_idx in group_indices:
```
Comment (LLMShark) [inline dance_grpo/dance_grpo_fsdp/diffusion_workers.py:642]: advantage的计算放一个函数里

File: dance_grpo/dance_grpo_fsdp/diffusion_workers.py [hunk 13]
```diff
@@ -0,0 +1,1380 @@
...
+                        (1.0 - t_val.item()) ** 2 + 1e-12
+                    )
+                    prev_sample_mean = prev_sample_mean + 0.5 * (eta**2) * score_estimate * (t_next_val - t_val).item()
+
+                    # Gaussian log-prob: log N(x_{t+1} | prev_sample_mean, std_dev_t^2 * I)
+                    diff = mb_next.to(torch.float32) - prev_sample_mean
+                    var_t = (std_dev_t**2).clamp_min(1e-20)
+                    log_std = torch.log(std_dev_t.clamp_min(1e-20))
+                    log_2pi = torch.log(torch.tensor(2.0 * math.pi, device=device, dtype=torch.float32))
+                    new_log_probs = -(diff**2) / (2.0 * var_t) - log_std - 0.5 * log_2pi
```
Comment (LLMShark) [inline dance_grpo/dance_grpo_fsdp/diffusion_workers.py:793]: DanceGRPO算法可以封装到函数里

File: dance_grpo/dance_grpo_fsdp/diffusion_workers.py [hunk 14]
```diff
@@ -0,0 +1,1380 @@
...
+                    # KL between N(mean_policy, sigma) and N(mean_ref, sigma) in closed form:
+                    # KL = ||mean_policy - mean_ref||^2 / (2 sigma^2)
+                    # (ref model would need a separate forward; omit if kl_coeff == 0)
+                    if kl_coeff > 0.0 and self.ref_module_fsdp is not None:
+                        with torch.no_grad():
+                            with torch.autocast(device_type=device_name, dtype=torch.bfloat16):
+                                if use_train_cfg:
+                                    cond_ref = self.ref_module_fsdp(
+                                        hidden_states=mb_current,
+                                        timestep=mb_ts,
```
Comment (LLMShark) [inline dance_grpo/dance_grpo_fsdp/diffusion_workers.py:832]: 这块逻辑需要看下是否放在这里

File: dance_grpo/dance_grpo_fsdp/diffusion_workers.py [hunk 15]
```diff
@@ -0,0 +1,1380 @@
...
+                        if gnorm > _best_gnorm:
+                            _best_gnorm = gnorm
+                            _debug_param_name = pname
+                            _debug_param_before = (p.data.float().norm().item(), gnorm, p.dtype)
+
+            self.actor_optimizer.step()
+            self.actor_lr_scheduler.step()
+
+            # Check if parameters actually changed after the update.
+            if torch.distributed.get_rank() == 0 and _debug_param_name is not None:
```
Comment (LLMShark) [inline dance_grpo/dance_grpo_fsdp/diffusion_workers.py:932]: debug相关的代码删掉

File: dance_grpo/dance_grpo_fsdp/diffusion_workers.py [hunk 16]
```diff
@@ -0,0 +1,1380 @@
...
+                "train/kl_loss": total_kl_loss,
+                "train/lr": self.actor_lr_scheduler.get_last_lr()[0],
+                # avg across ALL timesteps (was last-step only)
+                "train/ratio_mean": avg_ratio_mean,
+                "train/ratio_max": avg_ratio_max,  # worst-case max across all timesteps
+                "train/ratio_min": avg_ratio_min,  # worst-case min across all timesteps
+                # ---- Debug diagnostics ---- #
+                "debug/grad_norm_raw": grad_norm_raw,
+                "debug/grad_clip_ratio": grad_clip_ratio,
+                "debug/avg_group_reward_std": avg_group_reward_std,
```
Comment (LLMShark) [inline dance_grpo/dance_grpo_fsdp/diffusion_workers.py:982]: debug的metric都删掉

File: dance_grpo/dance_grpo_fsdp/diffusion_workers.py [hunk 17]
```diff
@@ -0,0 +1,1380 @@
...
+            offload_fsdp_optimizer(self.actor_optimizer)
+        if torch.distributed.get_rank() == 0:
+            logger.info(f"Loaded checkpoint from local_path: {local_path}")
+
+    # ---------------------------------------------------------------------- #
+    #  Fixed-evaluation helpers                                               #
+    # ---------------------------------------------------------------------- #
+
+    @register(dispatch_mode=Dispatch.ONE_TO_ALL)
+    def init_fixed_eval(self, data: DataProto):
```
Comment (LLMShark) [inline dance_grpo/dance_grpo_fsdp/diffusion_workers.py:1102]: init_fixed_eval和eval的流程是否可以统一

File: dance_grpo/dance_grpo_fsdp/diffusion_workers.py [hunk 18]
```diff
@@ -0,0 +1,1380 @@
...
+        if not hasattr(self, "_fixed_eval_samples") or not self._fixed_eval_samples:
+            logger.debug("run_fixed_eval: not initialised, skipping.")
+            return DataProto()
+
+        rank = torch.distributed.get_rank() if torch.distributed.is_initialized() else 0
+        global_step = data.meta_info.get("global_step", 0)
+        experiment_name = data.meta_info.get("experiment_name", "dance_grpo")
+
+        # Only evaluate every 10 RL steps (step 0, 10, 20, …)
+        if global_step % 10 != 0:
```
Comment (LLMShark) [inline dance_grpo/dance_grpo_fsdp/diffusion_workers.py:1209]: trainer.fit里控制eval流程，不要在worker里判断这些

File: dance_grpo/dance_grpo_fsdp/dance_ray_trainer.py [hunk 19]
```diff
@@ -0,0 +1,616 @@
...
+                        gen_batch_output = self.actor_rollout_wg.generate_sequences(gen_batch_output)
+
+                    batch = gen_batch_output
+
+                    with marked_timer("update_actor", timing_raw, "red"):
+                        actor_output = self.actor_rollout_wg.update_actor(batch)
+
+                    # After the weight update, render one image with the fixed evaluation
+                    # inputs to track how generation quality improves across steps.
+                    if self._fixed_eval_enabled:
```
Comment (LLMShark) [inline dance_grpo/dance_grpo_fsdp/dance_ray_trainer.py:460]: eval流程，统一在这里控制

File: dance_grpo/dance_grpo_fsdp/dance_ray_trainer.py [hunk 20]
```diff
@@ -0,0 +1,616 @@
...
+                # Compute reward and advantage metrics from the gathered batch DataProto
+                # so that mean/min/max reflect all samples across all workers.
+                if batch is not None and "rewards" in batch.batch.keys():
+                    # already scaled at reward model output
+                    rewards = batch.batch["rewards"].float().cpu()
+                    grpo_size = self.config.actor_rollout_ref.rollout.n
+                    _batch_size = rewards.shape[0]
+                    advantages = torch.zeros_like(rewards)
+                    group_indices = torch.chunk(torch.arange(_batch_size), _batch_size // grpo_size)
+                    for _g in group_indices:
```
Comment (LLMShark) [inline dance_grpo/dance_grpo_fsdp/dance_ray_trainer.py:492]: advantage计算只需要一次实现

File: dance_grpo/dance_grpo_fsdp/rollout.py [hunk 21]
```diff
@@ -0,0 +1,417 @@
...
+                    width=w,
+                    device=device,
+                    init_same_noise=init_same_noise,
+                    return_img_only=True,
+                )
+                rank = torch.distributed.get_rank() if torch.distributed.is_initialized() else 0
+                global_step = getattr(self, "_current_global_step", 0)
+                experiment_name = getattr(self, "_current_experiment_name", "dance_grpo")
+                step_dir = os.path.join(output_dir, experiment_name, f"step_{global_step:06d}")
+                os.makedirs(step_dir, exist_ok=True)
```
Comment (LLMShark) [inline dance_grpo/dance_grpo_fsdp/rollout.py:319]: 可以用ai优化一版，写法太复杂了

File: dance_grpo/dance_grpo_fsdp/train_grpo_edit.py [hunk 22]
```diff
@@ -0,0 +1,1865 @@
...
+            transformer.config.axes_lens,
+            theta=10000,
+        )
+    return freqs_cis
+
+
+def run_sample_step(
+    args,
+    z,
+    progress_bar,
```
Comment (LLMShark) [inline dance_grpo/dance_grpo_fsdp/train_grpo_edit.py:224]: 这个文件要删掉

File: dance_grpo/dance_grpo_fsdp/train_grpo_edit.py [hunk 23]
```diff
@@ -0,0 +1,1865 @@
...
+    sigma_schedule,
+    transformer,
```
Comment (LLMShark) [inline dance_grpo/dance_grpo_fsdp/train_grpo_edit.py:226]: 排查一下，是否有别的冗余代码文件

**Review by wuxibin89 — APPROVED** [review]
